### ISSUE TITLE
Inconsistent precompiliation errors from Cloudy example's dependencies

### ISSUE THREAD

**Issue by tsj5** [CliMA/CalibrateEmulateSample.jl#131]
**Problem statement**

Running `]precompile` or `]build` in the `examples/Cloudy` environment precompiles all dependencies successfully; however, proceeding to run the example via `include("Cloudy_example.jl")` precompiles some dependencies *again*, which fails.

This isa current problem in the feature branch for PR #130. The environment (Project and Manifest) for the Cloudy example in that branch is *identical* to `master`, yet the example runs in `master` but  fails to build dependencies on the feature branch.

**Steps to reproduce**

(taken from buildkite)

Using Julia 1.6 at the repo root directory, the following runs on `master` but fails on branch `tsj/add-explict-rng` (from #130).

```
julia --color=yes --project -e '
using Pkg;
Pkg.instantiate()
Pkg.activate("examples/Cloudy")
Pkg.instantiate()
println("+++ Running Cloudy")
include("examples/Cloudy/Cloudy_example.jl")'
```

**Details**

The error message begins with
```
ERROR: LoadError: MethodError: no method matching static_loop(::Symbol, ::Int64, ::Int64, ::Missing)
Closest candidates are:
  static_loop(::Symbol, ::Int64, ::Int64, ::Int64) at /Users/tsj/.julia/packages/LoopVectorization/kVenK/src/reconstruct_loopset.jl:64
Stacktrace:
  [1] LoopVectorization.Loop(ls::LoopVectorization.LoopSet, ex::Expr, sym::Symbol, f::Int64, s::Int64, l::Missing, ub::Int64)
    @ LoopVectorization ~/.julia/packages/LoopVectorization/kVenK/src/reconstruct_loopset.jl:26
...
```

Googling this yields no results. LoopVectorization is a sub-dependency of DiffEqBase, which is used by Sundials and DifferentialEquations.

The fact that precompilation is repeated in a "warm" REPL (i.e., run `using Pkg; Pkg.activate("."); Pkg.precompile()` with success, then at the next prompt run `include("Cloudy_example.jl")`, which re-precompiles and fails) appears to be improper behavior and an open bug in Pkg.jl (JuliaLang/Pkg.jl#2761). Cloudy_example changes `LOAD_PATH`, but that shouldn't invalidate existing build artifacts.

**Negative results/speculation**
-  `]up`ing the environment pulls in a new version of Sundials (4.9.0 -> 4.9.1) and dependencies, but does not change the error. 
- I haven't been successful at reducing the problem to a MWE. `]add`ing the versions of Sundials and DifferentialEquations to a fresh environment builds successfully. 
- 72063cae83a94b654029620b3ac21fa7fb25b3bf attempts to fix things by manually forcing a build, but this doesn't fix the issue locally. It was committed in the hopes that it might yield useful behavior when run in the CI environment.
- Issue is unchanged when all code after `using Sundials` is removed. 
- Cloudy hasn't been updated in some time; it may be that versions of its dependencies are inconsistently pinned and the issue here is a forwards compatibility problem.

**Comment by odunbar** [comment]
@bielim I don't suppose you have seen any of these issues before?

**Comment by tsj5** [comment]
Added copy-and-paste steps to reproduce the issue, for the benefit of anyone looking at this -- thanks for any help!

**Comment by tsj5** [comment]
I've made progress on this issue: the Sundials dependency is a red herring; the issue is version incompatibilities with Cloudy introduced by additions to CES. These are resolved in CliMA/Cloudy.jl#74.

# Steps to reproduce

The issue is present on the feature branch for PR #130. Commenting out all code past the `using` statements in Cloudy_example.jl and GModel.jl preserves the issue, but, importantly, commenting out the CES imports and the 
```
prepend!(LOAD_PATH, [joinpath(@__DIR__, "..", "..")])
```
statement causes all imports and precompilation to succeed. 

Modifying `LOAD_PATH` is apparently causing versions of dependencies from the CES environment to be used in the Cloudy environment. A more robust way to let the examples use pre-release versions of CES is to explicitly add the CES package to the examples as a relative path, but doing so results in a dependency conflict:

<details>
<summary>Version incompatibility (refactor-Sample branch)</summary>

```
(Cloudy) pkg> dev "../.."
   Resolving package versions...
ERROR: Unsatisfiable requirements detected for package FillArrays [1a297f60]:
 FillArrays [1a297f60] log:
 ├─possible versions are: 0.2.0-0.13.0 or uninstalled
 ├─restricted by compatibility requirements with Distributions [31c24e10] to versions: 0.9.0-0.11.9
 │ └─Distributions [31c24e10] log:
 │   ├─possible versions are: 0.16.0-0.25.48 or uninstalled
 │   ├─restricted to versions * by CalibrateEmulateSample [95e48a1f], leaving only versions 0.16.0-0.25.48
 │   │ └─CalibrateEmulateSample [95e48a1f] log:
 │   │   ├─possible versions are: 0.1.0 or uninstalled
 │   │   └─CalibrateEmulateSample [95e48a1f] is fixed to version 0.1.0
 │   ├─restricted by compatibility requirements with AdvancedMH [5b7e9947] to versions: 0.20.0-0.25.48
 │   │ └─AdvancedMH [5b7e9947] log:
 │   │   ├─possible versions are: 0.1.0-0.6.6 or uninstalled
 │   │   ├─restricted to versions * by CalibrateEmulateSample [95e48a1f], leaving only versions 0.1.0-0.6.6
 │   │   │ └─CalibrateEmulateSample [95e48a1f] log: see above
 │   │   └─restricted by compatibility requirements with Distributions [31c24e10] to versions: 0.5.4-0.6.6 or uninstalled, leaving only versions: 0.5.4-0.6.6
 │   │     └─Distributions [31c24e10] log: see above
 │   ├─restricted by compatibility requirements with GaussianProcesses [891a1506] to versions: 0.16.0-0.24.18, leaving only versions: 0.20.0-0.24.18
 │   │ └─GaussianProcesses [891a1506] log:
 │   │   ├─possible versions are: 0.8.0-0.12.4 or uninstalled
 │   │   ├─restricted to versions * by CalibrateEmulateSample [95e48a1f], leaving only versions 0.8.0-0.12.4
 │   │   │ └─CalibrateEmulateSample [95e48a1f] log: see above
 │   │   └─restricted by compatibility requirements with Distributions [31c24e10] to versions: 0.12.3-0.12.4 or uninstalled, leaving only versions: 0.12.3-0.12.4
 │   │     └─Distributions [31c24e10] log: see above
 │   └─restricted by compatibility requirements with EnsembleKalmanProcesses [aa8a2aa5] to versions: 0.24.14-0.25.48, leaving only versions: 0.24.14-0.24.18
 │     └─EnsembleKalmanProcesses [aa8a2aa5] log:
 │       ├─possible versions are: 0.1.0-0.3.0 or uninstalled
 │       └─restricted to versions * by CalibrateEmulateSample [95e48a1f], leaving only versions 0.1.0-0.3.0
 │         └─CalibrateEmulateSample [95e48a1f] log: see above
 └─restricted by compatibility requirements with Optim [429524aa] to versions: 0.6.2-0.8.14 — no versions left
   └─Optim [429524aa] log:
     ├─possible versions are: 0.15.3-1.6.1 or uninstalled
     ├─restricted by compatibility requirements with Cloudy [9e3b23bb] to versions: 0.19.4-0.20.6
     │ └─Cloudy [9e3b23bb] log:
     │   ├─possible versions are: 0.1.0 or uninstalled
     │   └─restricted to versions * by an explicit requirement, leaving only versions 0.1.0
     └─restricted by compatibility requirements with StatsBase [2913bbd2] to versions: [0.15.3, 0.20.5-1.6.1] or uninstalled, leaving only versions: 0.20.5-0.20.6
       └─StatsBase [2913bbd2] log:
         ├─possible versions are: 0.24.0-0.33.16 or uninstalled
         ├─restricted to versions * by CalibrateEmulateSample [95e48a1f], leaving only versions 0.24.0-0.33.16
         │ └─CalibrateEmulateSample [95e48a1f] log: see above
         ├─restricted by compatibility requirements with AbstractMCMC [80f14c24] to versions: 0.32.0-0.33.16
         │ └─AbstractMCMC [80f14c24] log:
         │   ├─possible versions are: 0.1.0-3.2.1 or uninstalled
         │   ├─restricted to versions * by CalibrateEmulateSample [95e48a1f], leaving only versions 0.1.0-3.2.1
         │   │ └─CalibrateEmulateSample [95e48a1f] log: see above
         │   ├─restricted by compatibility requirements with AdvancedMH [5b7e9947] to versions: [0.1.0, 0.3.0-3.2.1]
         │   │ └─AdvancedMH [5b7e9947] log: see above
         │   └─restricted by compatibility requirements with StatsBase [2913bbd2] to versions: 0.5.6-3.2.1 or uninstalled, leaving only versions: 0.5.6-3.2.1
         │     └─StatsBase [2913bbd2] log: see above
         └─restricted by compatibility requirements with EnsembleKalmanProcesses [aa8a2aa5] to versions: 0.33.0-0.33.16
           └─EnsembleKalmanProcesses [aa8a2aa5] log: see above
```
</details>

As shown above, the first incompatibility encountered is with FillArrays. The addition of AbstractMCMC/AdvancedMH in the feature branch requires recent versions of FillArrays, while the version of Optim used by Cloudy (0.20.6; current version of Optim is 1.6.1) requires older versions, resulting in an incompatibility. The AbstractMCMC/AdvancedMH dependencies aren't present in `master`, explaining why this issue isn't present on that branch despite its use of the same Project and Manifest for the Cloudy example.

# Proposed fix

Update dependencies used by Cloudy.jl; this is done in CliMA/Cloudy.jl#74.
